lib/keycard/institution_finder.rb

```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+require 'ipaddr'
+
+module Keycard
+  # looks up institution ID(s) by IP address
+  class InstitutionFinder
+    INST_QUERY = <<~SQL
+        SELECT inst FROM aa_network WHERE
+                ? >= dlpsAddressStart
+            AND ? <= dlpsAddressEnd
+            AND dlpsAccessSwitch = 'allow'
+            AND dlpsDeleted = 'f'
+            AND inst is not null
+      AND inst NOT IN
+          ( SELECT inst FROM aa_network WHERE
+            ? >= dlpsAddressStart
+            AND ? <= dlpsAddressEnd
+            AND dlpsAccessSwitch = 'deny'
+            AND dlpsDeleted = 'f' )
+    SQL
+
+    def initialize(db)
+      @db = db
+      @stmt = @db[INST_QUERY, *[:$client_ip] * 4].prepare(:select, :unused)
+    end
+
+    def attributes_for(request)
+      return {} unless (numeric_ip = numeric_ip(client_ip(request)))
+
+      insts = insts_for_ip(numeric_ip)
+
+      if !insts.empty?
+        { 'dlpsInstitutionIds' => insts }
```

Comment:
I'm not sure whether we should export the old names at this level (or come up with preferred naming/conventions here), but I think it's fine to start with this to get the plumbing in place.